File: 2024/python/day13.py
class Point:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y)

    def __le__(self, other):
        return self.x <= other.x and self.y <= other.y

    def __lt__(self, other):
        return self.x < other.x and self.y < other.y

    def __ge__(self, other):
        return self.x >= other.x and self.y >= other.y

    def __gt__(self, other):
        return self.x > other.x and self.y > other.y

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __ne__(self, other):
        return self.x != other.x or self.y != other.y

    def __repr__(self) -> str:
        return f"Point({self.x}, {self.y})"

    def copy(self):
        return Point(self.x, self.y)


def part1() -> int:
    machines: list[tuple[Point, Point, Point]] = []
    with open("../data/input13.txt") as inp:
        but_a: Point = Point()
        but_b: Point = Point()
        prize: Point = Point()
        for line in inp.readlines():
            if not line.strip():
                machines.append((but_a.copy(), but_b.copy(), prize.copy()))
            elif "Button A" in line:
                xy = line.split(":")[1].strip()
                but_a.x = int(xy.split(", ")[0].split("+")[1])
                but_a.y = int(xy.split(", ")[1].split("+")[1])
            elif "Button B" in line:
                xy = line.split(":")[1].strip()
                but_b.x = int(xy.split(", ")[0].split("+")[1])
                but_b.y = int(xy.split(", ")[1].split("+")[1])
            elif "Prize" in line:
                xy = line.split(":")[1].strip()
                prize.x = int(xy.split(", ")[0].split("=")[1])
                prize.y = int(xy.split(", ")[1].split("=")[1])

        machines.append((but_a.copy(), but_b.copy(), prize.copy()))

    cost = 0

    for a, b, p in machines:
        # print("Doing", a, b, p)
        min_total_cost = float("inf")
        best_a = best_b = 0
        curr = Point(0, 0)

        for bp in range(100):
            curr_b = Point(b.x * bp, b.y * bp)
            for ap in range(100):
                curr = curr_b + Point(a.x * ap, a.y * ap)
                if curr == p:
                    total_cost = ap * 3 + bp * 1
                    if total_cost < min_total_cost:
                        min_total_cost = total_cost
                        best_a = ap
                        best_b = bp

        cost += min_total_cost if min_total_cost != float("inf") else 0

    return int(cost)


def part2() -> int:
    machines: list[tuple[Point, Point, Point]] = []
    with open("../data/input13.txt") as inp:
        but_a: Point = Point()
        but_b: Point = Point()
        prize: Point = Point()
        for line in inp.readlines():
            if not line.strip():
                machines.append((but_a.copy(), but_b.copy(), prize.copy()))
            elif "Button A" in line:
                xy = line.split(":")[1].strip()
                but_a.x = int(xy.split(", ")[0].split("+")[1])
                but_a.y = int(xy.split(", ")[1].split("+")[1])
            elif "Button B" in line:
                xy = line.split(":")[1].strip()
                but_b.x = int(xy.split(", ")[0].split("+")[1])
                but_b.y = int(xy.split(", ")[1].split("+")[1])
            elif "Prize" in line:
                xy = line.split(":")[1].strip()
                prize.x = int(xy.split(", ")[0].split("=")[1])
                prize.y = int(xy.split(", ")[1].split("=")[1])
                prize.x += 10000000000000
                prize.y += 10000000000000

        machines.append((but_a.copy(), but_b.copy(), prize.copy()))

    cost = 0

    for a, b, p in machines:
        # print("Doing", a, b, p)
        min_total_cost = float("inf")
        best_a = best_b = 0
        curr = Point(0, 0)

        for bp in range(1000):
            curr_b = Point(b.x * bp, b.y * bp)
            print(bp)
            for ap in range(1000):
                curr = curr_b + Point(a.x * ap, a.y * ap)
                if curr == p:
                    total_cost = ap * 3 + bp * 1
                    if total_cost < min_total_cost:
                        min_total_cost = total_cost
                        best_a = ap
                        best_b = bp

        cost += min_total_cost if min_total_cost != float("inf") else 0

    return int(cost)


# print(part1())
print(part2())

File: 2024/cpp/day13.cpp
#include <climits>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;
using Point = pair<long int, long int>;

int part1();
int part2();

int main() {
  cout << part1() << endl;
  cout << part2() << endl;
}

int part1() {
  string line;
  ifstream input("../data/input13.txt");

  vector<vector<Point>> machines;

  if (input.is_open()) {

    vector<Point> machine;
    Point a;
    Point b;
    Point t;
    while (getline(input, line)) {
      if (line.empty()) {
        machine.push_back(a);
        machine.push_back(b);
        machine.push_back(t);
        machines.push_back(machine);
        machine.clear();
        continue;
      }

      if (line.find("Button A") != string::npos) {
        string temp, part;
        istringstream ss(line);
        getline(ss, temp, ':');
        getline(ss, part);

        string x, y;
        ss.clear();
        ss.str(part);
        getline(ss, x, ',');
        getline(ss, y);

        ss.clear();
        ss.str(x);
        getline(ss, temp, '+');
        getline(ss, x, '+');

        ss.clear();
        ss.str(y);
        getline(ss, temp, '+');
        getline(ss, y, '+');

        a.first = stoi(x);
        a.second = stoi(y);

      } else if (line.find("Button B") != string::npos) {
        string temp, part;
        istringstream ss(line);
        getline(ss, temp, ':');
        getline(ss, part);

        string x, y;
        ss.clear();
        ss.str(part);
        getline(ss, x, ',');
        getline(ss, y);

        ss.clear();
        ss.str(x);
        getline(ss, temp, '+');
        getline(ss, x, '+');

        ss.clear();
        ss.str(y);
        getline(ss, temp, '+');
        getline(ss, y, '+');

        b.first = stoi(x);
        b.second = stoi(y);

      } else if (line.find("Prize") != string::npos) {
        string temp, part;
        istringstream ss(line);
        getline(ss, temp, ':');
        getline(ss, part);

        string x, y;
        ss.clear();
        ss.str(part);
        getline(ss, x, ',');
        getline(ss, y);

        ss.clear();
        ss.str(x);
        getline(ss, temp, '=');
        getline(ss, x, '=');

        ss.clear();
        ss.str(y);
        getline(ss, temp, '=');
        getline(ss, y, '=');

        t.first = stoi(x);
        t.second = stoi(y);
      }
    }

    machine.push_back(a);
    machine.push_back(b);
    machine.push_back(t);
    machines.push_back(machine);
    machine.clear();
    input.close();
  }

  int cost = 0;

  for (const auto &m : machines) {
    Point a = m[0];
    Point b = m[1];
    Point p = m[2];

    int min_cost = INT_MAX;
    int best_a = 0, best_b = 0;
    int curr_x = 0, curr_y = 0;

    for (int bp = 0; bp < 100; bp++) {
      int curr_b_x = b.first * bp;
      int curr_b_y = b.second * bp;

      for (int ap = 0; ap < 100; ap++) {
        curr_x = curr_b_x + a.first * ap;
        curr_y = curr_b_y + a.second * ap;

        if (curr_x == p.first && curr_y == p.second) {
          int total_cost = ap * 3 + bp * 1;
          if (total_cost < min_cost) {
            min_cost = total_cost;
            best_a = ap;
            best_b = bp;
          }
        }
      }
    }

    if (min_cost != INT_MAX) {
      cost += min_cost;
    }
  }

  return cost;
}

int part2() { return 0; }
